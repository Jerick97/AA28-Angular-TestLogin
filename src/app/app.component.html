<div class="mx-auto mt-6 container">
    <span class="text-center font-bold text-xl">{{ title }} app is running!</span>
  
    <form [formGroup]="formLogin" (submit)="send()" autocomplete="off" class="container w-full max-w-xl p-8 mt-6 mx-auto space-y-6 rounded-md shadow bg-gray-900 ng-untouched ng-pristine ng-valid">
          <h2 class="w-full text-3xl text-white font-bold leading-tight">Login</h2>
          <div>
              <label for="username" class="block mb-1 ml-1 text-white">Username</label>
              <input formControlName="username" [ngClass]="{'correct': formLogin.get('username')?.valid,'error': formLogin.get('username')?.invalid && formLogin.get('username')?.touched}" id="username" type="text" placeholder="Username" class="block w-full text-white p-2 rounded focus:outline-none focus:ring focus:ring-opacity-25 focus:ring-violet-400 bg-gray-800">
              <div>
                  <p *ngIf="formLogin.get('username')?.errors?.['required']&& formLogin.get('username')?.touched" class="text-center block mb-1 ml-1 text-red-600">El Campo Username es Obligatorio</p>
                  <p *ngIf="formLogin.get('username')?.errors?.['minlength']&& formLogin.get('username')?.touched" class="text-center block mb-1 ml-1 text-red-600">Debe Tener como mínimo {{formLogin.get('username')?.errors?.['minlength']?.requiredLength}} caracteres</p>
                  <p *ngIf="formLogin.get('username')?.touched && formLogin.get('username')?.errors?.['pattern']" class="text-center block mb-1 ml-1 text-red-600">Formato de Username es Incorrecto email@dominio.com</p>
                  <p *ngIf="formLogin.get('username')?.valid" class="text-center block mb-1 ml-1 text-green-500">Todo Correcto</p>
              </div>
          </div>
          <div>
              <label for="password" class="block mb-1 ml-1 text-white">Password</label>
              <input formControlName="password" [ngClass]="{'correct': formLogin.get('password')?.valid,'error': formLogin.get('password')?.invalid && formLogin.get('password')?.touched}" id="password" type="password" placeholder="Contraseña" class="block w-full text-white p-2 rounded focus:outline-none focus:ring focus:ring-opacity-25 focus:ring-violet-400 bg-gray-800">
              <div>
                  <p *ngIf="formLogin.get('password')?.touched && formLogin.get('password')?.errors?.['required']" class="text-center block mb-1 ml-1 text-red-600">El Password es Obligatorio</p>
                  <p *ngIf="formLogin.get('password')?.errors?.['minlength']&& formLogin.get('password')?.touched" class="text-center block mb-1 ml-1 text-red-600">Debe Tener como mínimo {{formLogin.get('password')?.errors?.['minlength']?.requiredLength}} caracteres</p>
                  <p *ngIf="formLogin.get('password')?.valid" class="text-center block mb-1 ml-1 text-green-500">Todo Correcto</p>
              </div>
          </div>
          
          <div>
              <button [disabled]="formLogin.invalid" type="submit" class="btn-enviar w-full px-4 py-2 font-bold rounded shadow focus:outline-none focus:ring hover:ring focus:ring-opacity-50 bg-violet-400 focus:ring-violet-400 hover:ring-violet-400 text-gray-900">Confirmar</button>
          </div>
          <div *ngIf="formLogin.valid && correcto">
              <p class="text-center block mb-1 ml-1 text-green-500">Mensaje Enviado Correctamente</p>
          </div>
          <div class="text-center">
              <p class="block mb-1 ml-1 text-white font-bold p-4">Form Invalid - <span [ngClass]="{'bg-red-500':formLogin.invalid == false, 'bg-green-500' : formLogin.invalid == true}" class="rounded-full text-white px-3 py-1">{{formLogin.invalid}}</span></p>
              <p class="block mb-1 ml-1 text-white font-bold p-4">Form Valid - <span [ngClass]="{'bg-red-500':formLogin.valid == false, 'bg-green-500' : formLogin.valid == true}" class="rounded-full text-white px-3 py-1">{{formLogin.valid}}</span></p>
          </div>
          <div class="bg-black p-4 text-1xl text-yellow-400 mt-4">
              <p>Username : {{formLogin.get('username')?.value}}</p>
              <p>Password : {{formLogin.get('password')?.value}}</p>
          </div>
      </form>
  </div>
  <router-outlet></router-outlet>