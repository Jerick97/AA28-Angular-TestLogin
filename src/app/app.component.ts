import { Component, OnInit } from '@angular/core';
import { FormBuilder, FormGroup, Validators } from '@angular/forms';
@Component({
  selector: 'app-root',
  templateUrl: './app.component.html',
  styleUrls: ['./app.component.css']
})
export class AppComponent implements OnInit{
  title = 'AppLoginTest';
  correcto : boolean = false;
  public formLogin!: FormGroup;

  constructor(private formBuilder:FormBuilder) { }

  ngOnInit(): void {
    this.formLogin = this.formBuilder.group({
      username: ['', [
        Validators.required,
        Validators.minLength(4),
        Validators.pattern(/^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,4}$/)
      ]],
      password: ['', [
        Validators.required,
        Validators.minLength(4)
      ]],
      
    });
  }

  send(): any{
    console.log(this.formLogin.value);

    //DespuÃ©s de 3 Segundos, Reinicia el Form
    setTimeout(() => {
      this.correcto = false;
      this.formLogin.reset();
    }, 3000);

    this.correcto = true;
  }
}
