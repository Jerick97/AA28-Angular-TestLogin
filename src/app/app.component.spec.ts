import { TestBed } from '@angular/core/testing';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { By } from '@angular/platform-browser';
import { RouterTestingModule } from '@angular/router/testing';
import { AppComponent } from './app.component';

describe('AppComponent', () => {
  beforeEach(async () => {
    await TestBed.configureTestingModule({
      imports: [
        RouterTestingModule,
        FormsModule,
        ReactiveFormsModule
      ],
      declarations: [
        AppComponent
      ],
    }).compileComponents();
  });

  it('should create the app', () => {
    const fixture = TestBed.createComponent(AppComponent);
    const component = fixture.componentInstance;
    expect(component).toBeTruthy();
  });

  it(`should have as title 'AppLoginTest'`, () => {
    const fixture = TestBed.createComponent(AppComponent);
    const component = fixture.componentInstance;
    expect(component.title).toEqual('AppLoginTest');
  });

  it('should render title', () => {
    const fixture = TestBed.createComponent(AppComponent);
    fixture.detectChanges();
    const compiled = fixture.nativeElement as HTMLElement;
    expect(compiled.querySelector('span')?.textContent).toContain('AppLoginTest app is running!');
  });

  //La siguiente prueba tomara una captura del código, si mas adelante alguien modificara el codigo la prueba arrojaría un error mostrando el cambio que se realizo
  it('Se espera hacer match con el snapshot', () => {
    const fixture = TestBed.createComponent(AppComponent);
    fixture.detectChanges(); //para capturar el texto de las variables y no solo la estructura html
    const compiled = fixture.nativeElement as HTMLElement;
    expect(compiled).toMatchSnapshot(); //esto creara una copia del app.component.html en una carpeta y lo compara con el código actual
  })

  it('Se esperaba formLogin.valid retorne verdadero cuando los valores de username y password sean correcto', () => {
    const fixture = TestBed.createComponent(AppComponent);
    const app = fixture.componentInstance;
    fixture.detectChanges();

    app.formLogin.controls['username'].setValue('emerson@gmail.com');
    app.formLogin.controls['password'].setValue('12345678');
    expect(app.formLogin.valid).toBeTruthy();
    });
    
  it('Se esperaba formLogin.invalid retorne verdadero cuando los valores de username o password sean incorrecto', () => {
    const fixture = TestBed.createComponent(AppComponent);
    const app = fixture.componentInstance;
    fixture.detectChanges();
    app.formLogin.controls['username'].setValue('emerson');
    app.formLogin.controls['password'].setValue('');
    expect(app.formLogin.invalid).toBeTruthy();
  });

  it('Se esperaba formLogin.invalid al inicio retorne falso', () => {
    const fixture = TestBed.createComponent(AppComponent);
    const app = fixture.componentInstance;
    fixture.detectChanges();

    app.formLogin.controls['username'];
    app.formLogin.controls['password'];
    expect(app.formLogin.invalid).toBeTruthy();
    });

    it('Se espera que la variable correcto sea true al pulsar el boton y luego de 3 segundos false', () => {
      jest.useFakeTimers();
      const fixture = TestBed.createComponent(AppComponent);
      const component = fixture.componentInstance;
      fixture.detectChanges();
      component.formLogin.controls['username'].setValue('emerson@gmail.com');
      component.formLogin.controls['password'].setValue('12345678');
      const miBoton = fixture.debugElement.query(By.css('.btn-enviar'));
      
      miBoton.nativeElement.click(); //si se ingresa los input correctamente ejecutamos el click en el button
      expect(component.correcto == true); //validamos que la variable correcto sea verdadero antes de los 3 segundos
      jest.advanceTimersByTime(3000); //simulamos el setTimeout que está avanzando el tiempo en 3 segundos
      expect(component.correcto).toBeFalsy(); //validamos que la variable correcto sea falso antes de los 3 segundos
  });
});
