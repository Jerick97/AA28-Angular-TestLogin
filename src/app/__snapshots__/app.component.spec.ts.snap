// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppComponent Se espera hacer match con el snapshot 1`] = `
<div
  id="root3"
>
  <div
    class="mx-auto mt-6 container"
  >
    <span
      class="text-center font-bold text-xl"
    >
      AppLoginTest app is running!
    </span>
    <form
      autocomplete="off"
      class="container w-full max-w-xl p-8 mt-6 mx-auto space-y-6 rounded-md shadow bg-gray-900 ng-untouched ng-pristine ng-valid ng-invalid"
      novalidate=""
    >
      <h2
        class="w-full text-3xl text-white font-bold leading-tight"
      >
        Login
      </h2>
      <div>
        <label
          class="block mb-1 ml-1 text-white"
          for="username"
        >
          Username
        </label>
        <input
          class="block w-full text-white p-2 rounded focus:outline-none focus:ring focus:ring-opacity-25 focus:ring-violet-400 bg-gray-800 ng-untouched ng-pristine ng-invalid"
          formcontrolname="username"
          id="username"
          placeholder="Username"
          type="text"
        />
        <div>
          
          
          
          
        </div>
      </div>
      <div>
        <label
          class="block mb-1 ml-1 text-white"
          for="password"
        >
          Password
        </label>
        <input
          class="block w-full text-white p-2 rounded focus:outline-none focus:ring focus:ring-opacity-25 focus:ring-violet-400 bg-gray-800 ng-untouched ng-pristine ng-invalid"
          formcontrolname="password"
          id="password"
          placeholder="ContraseÃ±a"
          type="password"
        />
        <div>
          
          
          
        </div>
      </div>
      <div>
        <button
          class="btn-enviar w-full px-4 py-2 font-bold rounded shadow focus:outline-none focus:ring hover:ring focus:ring-opacity-50 bg-violet-400 focus:ring-violet-400 hover:ring-violet-400 text-gray-900"
          disabled=""
          type="submit"
        >
          Confirmar
        </button>
      </div>
      
      <div
        class="text-center"
      >
        <p
          class="block mb-1 ml-1 text-white font-bold p-4"
        >
          Form Invalid - 
          <span
            class="rounded-full text-white px-3 py-1 bg-green-500"
          >
            true
          </span>
        </p>
        <p
          class="block mb-1 ml-1 text-white font-bold p-4"
        >
          Form Valid - 
          <span
            class="rounded-full text-white px-3 py-1 bg-red-500"
          >
            false
          </span>
        </p>
      </div>
      <div
        class="bg-black p-4 text-1xl text-yellow-400 mt-4"
      >
        <p>
          Username : 
        </p>
        <p>
          Password : 
        </p>
      </div>
    </form>
  </div>
  <router-outlet />
  
</div>
`;
